_plugins/editable_docs.rb

```diff
@@ -0,0 +1,21 @@
+Jekyll::Hooks.register :site, :pre_render do |site, payload|
+  repositories = site.data["remotes"]["repositories"]
+  site.pages.each do |page|
+    page_path = page.path
+    if page.data.key?("origin_path")
+      page_path = page.data["origin_path"]
+    end
+    next unless page_path.start_with?("doc")
+    repo_name, repo_data = repositories.find do |name, data|
+      page_path.start_with? (File.join("doc", name))
+    end
+    next if repo_name.nil? or repo_data.nil?
+    next unless repo_data.key?("edit_url")
+    page_relative_url = page_path.sub(File.join("doc", repo_name), "")
+    # Path joins won't look past a potential trailing slash (i.e.
+    # won't care about one of the fragments being a partial URL).
+    page.data["edit_url"] = File.join(
+      repo_data["edit_url"], page_relative_url
```

Comment:
Instead of requiring encoding the edit_url into the repository I think we can make this "smarter" where we do the following logic. 

```
if url domain == 'github.com':
  Detect org, repo, branch from URL
  Substitute into github edit url template
if url domain == "bitbucket.org"
  Detect org, repo, branch from URL
  Substitute into bitbucket edit url template
```

With this we can add support for any public hosting repository and it will start working for any repo using that pattern. And we can extend it site wide with new providers that will retroactively apply to preexisting repos.

This can also be used for the readmes relatively easily on multiple platforms.

Reply:
@tfoote @apojomovsky I agree with both of you, not being explicit about the URL is the cleanest approach. It has a few caveats though (which may shed some light as to why I chose not to at first):

- AFAIK VCS version field may be a commit hash or a tag. Where should the edit link take you then? To master? To default?
- VCS URL field may be (and it is in our case) an ssh-like URL which carries no branch information and makes regular URI parsers throw.

Thus, I chose to be explicit rather than implicit. No funky rules to remember, plain wysiwyg. Plus you can just drop the field to disable the feature. 

We can, however, exercise a best effort approach and disable the feature when it's not possible to infer the target URL. What do you think?

Reply:
> AFAIK VCS version field may be a commit hash or a tag. Where should the edit link take you then? To master? To default?

It certainly can be. For the documentation stuff we should just standardize/require the use of a branch name and then we don't need to worry about these complexities.

> VCS URL field may be (and it is in our case) an ssh-like URL which carries no branch information and makes regular URI parsers throw.

Related to this we should standardize on https urls not ssh urls. Without this you cannot clone one a machine that doesn't have a github account. (You have to have your ssh key registered with github etc to be able to clone over ssh), which prevents anonymous access.

If we want to get fancy we can add linters for both these conditions. These are both policies we enforce in the rosdistro repository.